File: 2022/C++/Day-2/Part-2.cpp
#include <map>
#include <array>
#include <fstream>
#include <iostream>

int main() {
    std::ios_base::sync_with_stdio(false);
    std::ifstream inputFile("input.txt");
    //-------------------------------------
    int score(0);
    {
        std::map<std::string, int> outcomes = {
            {"A X", 3}, {"A Y", 6}, {"A Z", 0},
            {"B X", 0}, {"B Y", 3}, {"B Z", 6},
            {"C X", 6}, {"C Y", 0}, {"C Z", 3}
        };
        std::map<char, std::array<char, 3>> choices = {
            {'A', {'Y', 'X', 'Z'}}, 
            {'B', {'Z', 'Y', 'X'}}, 
            {'C', {'X', 'Z', 'Y'}} 
        };
        std::map<char, int> scores = {{'X', 1}, {'Y', 2}, {'Z', 3}};

        std::string round;
        while (std::getline(inputFile, round)) {
            if (round.empty()) break;
            switch (round[2]) {
                case 'X': round[2] = choices[round[0]][2]; break;
                case 'Y': round[2] = choices[round[0]][1]; break;
                case 'Z': round[2] = choices[round[0]][0]; break;
            }
            score += scores[round[2]] + outcomes[round];
        }
    }
    std::cout << score;
    //-------------------------------------
    return 0;
}

File: 2022/Python/Day-2/Part-2.py
from helper_functions import get_relative_path


outcomes = {
    "A X": 3, "B X": 0, "C X": 6,
    "A Y": 6, "B Y": 3, "C Y": 0,
    "A Z": 0, "B Z": 6, "C Z": 3}

choices = { 
    "A": ["Y", "X", "Z"], 
    "B": ["Z", "Y", "X"], 
    "C": ["X", "Z", "Y"]}

scores = {"X": 1, "Y": 2, "Z": 3}


def main(): 
    score = 0

    with open(get_relative_path("input.txt")) as file:
        while Round := file.readline().strip():
            match Round[2]:
                case "X":
                    Round = Round[:2] + choices[Round[0]][2]
                case "Y":
                    Round = Round[:2] + choices[Round[0]][1]
                case "Z":
                    Round = Round[:2] + choices[Round[0]][0]

            score += scores[Round[2]] + outcomes[Round]
            
    print(score)


if __name__ == "__main__":
    main()
